src/main/java/org/openstreetmap/atlas/geography/atlas/change/diff/AtlasDiff.java

```diff
@@ -0,0 +1,467 @@
+     */
+    public AtlasDiff useBloatedEntities(final boolean useBloatedEntities)
+    {
+        this.useBloatedEntities = useBloatedEntities;
+        return this;
+    }
+
+    /**
+     * Use geometry matching when computing diffs. This means that if we detect an added or removed
+     * {@link Edge} based on ID, we will additionally check its geometry before actually generating
+     * a diff. Sometimes, the same OSM way can be way-sectioned differently. So while the IDs for a
+     * sectioned OSM way will be different, the underlying geometry will be the same. In that case,
+     * {@link AtlasDiff#useGeometryMatching(boolean)} will cause {@link AtlasDiff} to ignore the
+     * inconsistent way sectioning.
+     *
+     * @param useGeometryMatching
+     *            use geometry matching
+     * @return a configured {@link AtlasDiff}
+     */
+    public AtlasDiff useGeometryMatching(final boolean useGeometryMatching)
+    {
+        this.useGeometryMatching = useGeometryMatching;
+        return this;
+    }
+
+    private Set<FeatureChange> createFeatureChangesBasedOnEntitySets(
+            final Set<AtlasEntity> addedEntities, final Set<AtlasEntity> removedEntities,
+            final Set<AtlasEntity> potentiallyModifiedEntities, final Atlas beforeAtlas,
+            final Atlas afterAtlas, final boolean useGeometryMatching,
+            final boolean useBloatedEntities, final boolean saveAllGeometries)
+    {
+        final Set<FeatureChange> featureChanges = new HashSet<>();
+
+        addedEntities.stream()
+                .map(addedEntity -> createSimpleFeatureChangeWithType(ChangeType.ADD, addedEntity,
+                        afterAtlas, useBloatedEntities, saveAllGeometries))
+                .forEach(featureChanges::add);
+
+        removedEntities.stream()
+                .map(removedEntity -> createSimpleFeatureChangeWithType(ChangeType.REMOVE,
+                        removedEntity, beforeAtlas, useBloatedEntities, saveAllGeometries))
+                .forEach(featureChanges::add);
+
+        potentiallyModifiedEntities.stream()
+                .map(modifiedEntity -> createModifyFeatureChanges(modifiedEntity, beforeAtlas,
+                        afterAtlas, useGeometryMatching, useBloatedEntities, saveAllGeometries))
+                .forEach(modifyFeatureChangeSet -> modifyFeatureChangeSet
+                        .forEach(featureChanges::add));
+
+        return featureChanges;
+    }
+
+    /**
+     * Create a set of modify {@link FeatureChange}s ie. ADDs that change an existing feature. The
+     * set will contain an individual {@link FeatureChange} for certain types of differences. Eg. if
+     * a Point's location AND tags changed, then the set will contain two feature changes, one for
+     * the location and one for the tags. However, if a Node's location and connected edges change,
+     * this will come in as one feature change.<br>
+     *
+     * @param entity
```

Comment:
We usually expect comments on each variable here if the javadoc comment is present.